update stac.py, create-item, 2020

pholleway opened the pull request:
**Proposed Changes:**

1. Update create-item (2020 section). The function extracts a str from an xml file for item id & dt. Previously, the function couldn't handle a str with ".tif" extension and would throw error. The code now checks for extension and removes it if present. 

This issue is due to various inconsistencies across fields in the NAIP 2020 xml metadata. Additionally, file structure is inconsistent. 

**PR Checklist:**

- [X] I have added my changes to the [CHANGELOG](https://github.com/stactools/naip/blob/main/CHANGELOG.md) **or** a CHANGELOG entry is not required.

philvarner commented on src/stactools/naip/stac.py:
```diff
@@ -158,7 +159,7 @@ def create_item(
                     )
                 if resource_desc is not None:
-                    dt = str_to_datetime(resource_desc.split("_")[-1])
+                    resource_desc, dt = process_resource_desc(resource_desc)
```
the signature of `process_resource_desc` is `Optional[Tuple[str, datetime]]`, so this will fail if it ever returns None, so you need to handle that case.

philvarner replied:
actually, maybe the signature on process_resource_desc doesn't match the domain of values returned by the method?

pholleway replied:
Adjusted function to only process the resource desc. The code for dt won't be duplicated and can be within stac.py

philvarner commented on src/stactools/naip/utils.py:
```diff
@@ -91,3 +92,15 @@ def maybe_extract_id_and_date(cog_href: str) -> Optional[Tuple[str, datetime]]:
     return resource_desc, dt
+
+
+def process_resource_desc(resource_desc: str) -> Optional[Tuple[str, datetime]]:
```
you have a check if `resource_desc is not None`, so that indicates the parameter signature should be `resource_desc: Optional[str]`

philvarner replied:
I would recommend splitting this method into two methods. It seems be be doing two unreleated things -- stripping the `.tif` extension from the resource_desc and determining the datetime. It's not ideal to have one method that returns a tuple with unrelated values like this.

philvarner replied:
As the method is written, the signature is not `Optional[Tuple[str, datetime]]` but rather `Tuple[Optional[str], Optional[datetime]]`, since you're always returning a Tuple, but the values in it may be None. (but this method should be split up anyway, so it doesn't matter)

pholleway replied:
Thanks for the feedback. 

Adjusted function to only process the resource desc. The code for dt won't be duplicated and can be within stac.py

philvarner reviewed (changes requested):
comments inline

pholleway commented:
@philvarner addressed your comments. Also, I noticed there were no tests capturing for resource_desc extracted from the cog_href. 

"resource_desc = root.find_text_or_throw(second_xpath, missing_element("File Identifier")" will cause tests to fail, which is incorrect. The code continues and attempts to find resource_desc and dt from the cog_href. 

I modified this code to follow years <2020, since there is no exception while searching the metadata. The exception is raised once all methods fail.

philvarner commented on tests/data-files/m_3211605_ne_11_060_20200415_20200730.xml:
```diff
@@ -5,7 +5,7 @@
 			<citeinfo>
 				<origin>USDA-Farm Production and Conservation Business Center (FPAC-BC)</origin>
 				<pubdate>20200729</pubdate>
-				<title>m_3211605_ne_11_060_20200415.tif</title>
```
was this intentionally removed?

pholleway replied:
no, thanks. It was removed whenever testing the 2020 resource_desc from cog_href. 

Added back in. There was an additional file representing this scenario committed.

philvarner commented on tests/data-files/m_3211605_ne_11_060_20200415_20200730.xml:
```diff
@@ -0,0 +1,221 @@
+			<citeinfo>
+				<origin>USDA-Farm Production and Conservation Business Center (FPAC-BC)</origin>
+				<pubdate>20200729</pubdate>
+				<title></title>
```
these two xml files are the same. Is this one supposed to have the title value set?

pholleway replied:
Added to have create item pull resource_desc and dt from cog_href (not xml). This wasn't included in tests and did throw an error whenever tested.

philvarner commented on tests/test_stac.py:
```diff
@@ -91,3 +91,35 @@ def test_create_item_xml_grid_code(self):
+
+    # test stac on year 2020
+    # handles resource description with and without ".tif" extension
+    def test_create_item_xml_ext(self):
```
these two tests are doing the same operation on two different files and expecting the same results. They could be collapsed into one test, and looped like:

``
for filename in [
"data-files/m_3211605_ne_11_060_20200415_20200730.xml",
"data-files/m_3211605_ne_11_060_20200415_20200730_missing_resource_desc.xml"]:
  item = create_item(
            "ca",
            "2020",
            test_data.get_path("data-files/m_3211605_ne_11_060_20200415.tif"),
            test_data.get_path(filename),
        )
   ...
```

philvarner commented on tests/test_stac.py:
```diff
@@ -91,3 +91,35 @@ def test_create_item_xml_grid_code(self):
+
+        self.assertEqual(item.id, "ca_m_3211605_ne_11_060_20200415")
+        self.assertEqual(
+            item.datetime, str_to_datetime("20200415") + timedelta(hours=16)
```
Why isn't this comparing an exact datetime value instead of constructing it from a string and timedelta?

pholleway replied:
item.datetime was originally constructed as dt = dt + timedelta(hours=16)

- An annotation states "UTC is +4 ET". I believe the 16 timedelta deals with the conversion between 1) military/standard (+12) and 2) UTC/ET (+4). 

- The purpose of the time delta is a bit unclear, since it doesn't seem like NAIP reports actual times (based on spot checks on a few items). Time stamps are generated from the date and therefore always 00:00:00. The timedelta conversion doesn't really seem to provide any meaningful. 

Let me know your thoughts.

philvarner replied:
I think you can just to this, which is more clear: str_to_datetime("2020-04-15T16:00:00Z")

pholleway replied:
updated  to -> datetime.strptime("2020-09-02T16:00:00Z", "%Y-%m-%dT%H:%M:%SZ")

assert equal failing with -> str_to_datetime("2020-04-15T16:00:00Z")

philvarner commented on tests/test_stac.py:
```diff
@@ -74,7 +73,8 @@ def test_incorrect_metadata_txt(self):
         self.assertEqual(
-            item.datetime, str_to_datetime("20110815") + timedelta(hours=16)
+            item.datetime,
+            datetime.strptime("2011-08-15T16:00:00Z", "%Y-%m-%dT%H:%M:%SZ"),
```
For future reference: https://docs.python.org/3.8/library/datetime.html#datetime.datetime.fromisoformat
(which on python < 3.11 (or 3.10?) requires a timezone of `+00:00` instead of `Z`
